Source/Application/Use Cases/Restaurant List/Controller/RestaurantListController.swift

```diff
@@ -0,0 +1,73 @@
+import UIKit
+
+final class RestaurantListController: NSObject {
+
+    private static let listCellReuseIdentifier = "_restaurant_list_cell"
+    private let dependencies: HasImageProvider
+
+    private let restaurants: [Restaurant]
+
+    private var images: [Restaurant: UIImage] = Dictionary()
+    private var isLoadingImage: [Restaurant: Bool] = Dictionary()
+
+    init(dependencies: HasImageProvider, restaurants: [Restaurant]) {
+        self.dependencies = dependencies
+        self.restaurants = restaurants
+    }
+
+    func prepare(tableView: UITableView) {
+        tableView.register(RestaurantListCell.self, forCellReuseIdentifier: RestaurantListController.listCellReuseIdentifier)
+    }
+}
+
+extension RestaurantListController: UITableViewDataSource {
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return self.restaurants.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: RestaurantListController.listCellReuseIdentifier, for: indexPath)
+        if let cell = cell as? RestaurantListCell {
+
+            let restaurant = self.restaurants[indexPath.row]
+
+            cell.update(with: RestaurantListCellViewModel(restaurant: restaurant))
+
+            if let image = self.images[restaurant] {
```

Comment:
I feel like this may cause issues and raise potential race conditions (in an unlikely event), in theory. 

Additionally, I feel like we can test the logic that is happening here. I feel that there are indeed two concerns that we can extract out of our Controller here (much like the intent of our "Coordinator" objects!).

1. The (in-memory) caching of our images, and
2. The logic to decide to perform the network load.

What do you think about that work?